contracts/utils/math/Math.sol

```diff
@@ -537,41 +537,82 @@ library Math {
+        //    to the next power of 2` algorithm, the version used here is modified to round down instead of up.
+        //    ref: https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+        //
+        // 2. Compute `n ≡ x mod 255`, we use this `n` value to split the equation in two parts that can be more
+        //    efficiently computed using the `Logarithm's Product Rule`:
+        //      - lemma: log2(x) == log2(x / n) + log2(n)
+        //    this specific `n` allows a very efficient lookup table to be used to calculate both `log2(x / n)` and `log2(n)`.
+        //
+        // 3. Compute `prod0 = log2(n)`, once `x` is a power of two, there are only 8 possible values for `n`, so `log2(n)`
+        //    can be trivially computed using a lookup table, here we use the opcode `BYTE` to lookup a 256bit word.
+        //
+        // 4. Compute `prod1 = log2(x / 2**prod0)`, the divisor `2**prod0` guarantees the resulting value is always multiple
+        //    of 8, there's exactly 32 distinct values for `prod1`, respectively: 0, 8, 16, .. , 248. This allow an efficient
+        //    lookup table to be created using a single word, and multiplication and shift extract the value.
+        //
+        // 5. The final result is simply the sum of `prod0` and `prod1` calculated previously:
+        //    floor(log2(x)) = prod0 + prod1
+        //
+        // @author Lohann Ferreira <developer@lohann.dev>
         unchecked {
-            exp = 128 * SafeCast.toUint(value > (1 << 128) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 64 * SafeCast.toUint(value > (1 << 64) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 32 * SafeCast.toUint(value > (1 << 32) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 16 * SafeCast.toUint(value > (1 << 16) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 8 * SafeCast.toUint(value > (1 << 8) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 4 * SafeCast.toUint(value > (1 << 4) - 1);
-            value >>= exp;
-            result += exp;
-
-            exp = 2 * SafeCast.toUint(value > (1 << 2) - 1);
-            value >>= exp;
-            result += exp;
+            // Round `x` down to the closest power of 2 using the Seander's algorithm.
+            // Reference: https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+            x |= x >> 1;
+            x |= x >> 2;
+            x |= x >> 4;
+            x |= x >> 8;
+            x |= x >> 16;
+            x |= x >> 32;
+            x |= x >> 64;
+            x |= x >> 128;
+            x = (x >> 1) + 1;
+
+            uint256 prod0;
+            assembly {
+                // 1. Compute `n ≡ x mod 255`, because `x` is power of two, the resulting `n` can only be one of the
```

Comment:
I'd write it as
```suggestion
                // 1. Compute `x ≡ n mod 255`, because `x` is power of two, the resulting `n` can only be one of the
```
or
```suggestion
                // 1. Compute the remainder of `x` modulo 255, because `x` is power of two, the resulting `n` can only be one of the
```